### PR TITLE
add SSP to tokenizers

### PR DESCRIPTION
**Pull request by henchc**
Addresses one part of [#2203](https://github.com/nltk/nltk/issues/2203). If this can get refined and merged I can work on the Legality Principle tokenizer.

### CODE REVIEW

**Comment by alvations** [comment]
Thank you for the suggestion @no-identd. Thank you in advance for contributing too, @henchc 

Since it looks like it can be useful but we lack knowledge of how it works,  what the limitations are and how to maintain it if this is merged into NLTK.

So I've just a few question to ask before we proceed to review. 

   - Am I right to say that the current implementation in the PR only supports English syllabification of word?

   - What is the intended input to the `tokenize()` function? A sentence string? Or a single word. If it's a sentence string, how does it differentiate between the word and the syllable boundaries? 

   - If we want to support more languages, where do we find a list of `sonority_hierarchy` for other languages? 

   - What are the languages that the principle covers? Only languages with romanizable alphabets? Or would it also work on something like Chinese pinyin? 

   - Is space necessary for the implementation to work? For example, the algorithm is given romanized Japanese without spaces would it still work? 

  - Is there a canonical reference publication that shows that the  `Sonority Sequencing Principle` works for languages outside of German? Also, is there a specific chapter from ["Lehrbuch der Phonetik."](https://ia902609.us.archive.org/23/items/lehrbuchderphone00jespuoft/lehrbuchderphone00jespuoft_bw.pdf) that users/readers of the code should refer to to understand how it works? Citing the whole 200+ pages textbook might not be optimal for something that isn't mainstream. 

Sorry in advance for the multiple question. They are asked so that we better understand what we merge to the codebase and document it in the code docstring with more specificity. And we know how to maintain or extend it, if needed.

**Comment by henchc** [comment]
Thanks @alvations ! And no worries, I'm happy to document whatever is necessary.

- Am I right to say that the current implementation in the PR only supports English syllabification of word?

> The default hierarchy in the PR will only support English, and not as fine-grained as an IPA translation would. The SSP is a universal syllabification algorithm, but that doesn't mean it performs equally across languages. [Wikipedia](https://en.wikipedia.org/wiki/Sonority_Sequencing_Principle) actually has a really good summary. Relatedly, [Bartlett et al.](http://www.anthology.aclweb.org/N/N09/N09-1035.pdf) is a good benchmark for English accuracy if utilizing IPA (see their hierarchy on pg. 311). The idea with this PR is to allow users to provide the hierarchy for whatever language their working with. The implementation @no-identd first found was my implementation for medieval German. This is definitely more targeted for NLP researchers as opposed to an out-of-the-box solution for production syllabification. Though as the paper shows, the accuracy is quite high from this algorithm alone provided an accurate hierarchy.

- What is the intended input to the tokenize() function? A sentence string? Or a single word. If it's a sentence string, how does it differentiate between the word and the syllable boundaries?

> Intended input is any string the user wishes to be broken into syllables. For word-based languages this would be a word. For other languages it could be any sequence of phonemes. So it does not differentiate between word boundaries, it simply takes a string *without* spaces and will syllabify it. For logographic/ideographic languages this would need to be translated out of the script first. 

- If we want to support more languages, where do we find a list of `sonority_hierarchy` for other languages?

> Good question! I'm not sure of a resource with all language hierarchies. My suggested approach would be to go to the wikipedia page for the phonology of a language, e.g. [Turkish](https://en.wikipedia.org/wiki/Turkish_phonology) and then create the hierarchy from the consonant and vowel tables. As the [Bartlett et al.](http://www.anthology.aclweb.org/N/N09/N09-1035.pdf) paper discovers (as I did in my research), a strict and large hierarchy is not always most accurate. It's sometimes better to collapse certain levels.

- What are the languages that the principle covers? Only languages with romanizable alphabets? Or would it also work on something like Chinese pinyin?

> The accuracy varies by language. Generally, if the script is converted to phonemes in IPA and an IPA hierarchy is used for the language it can be quite accurate for most languages. As stated above, logographic/ideographic scripts would have to be converted as they are not phoneme based to any degree (English, e.g., is also not a 1-1 orthography to phonology, but the given hierarchy approximates it).

- Is space necessary for the implementation to work? For example, the algorithm is given romanized Japanese without spaces would it still work?

 > See input response above, but no space should be included in the input to `.tokenize()`

- Is there a canonical reference publication that shows that the Sonority Sequencing Principle works for languages outside of German? Also, is there a specific chapter from "Lehrbuch der Phonetik." that users/readers of the code should refer to to understand how it works? Citing the whole 200+ pages textbook might not be optimal for something that isn't mainstream.

> I tried citing the earliest reference I am aware of, but the principle is widely discussed in the literature. A common citation for its explanation is Selkirk: *Selkirk, E. (1984). On the major class features and syllable theory. In Aronoff & Oehrle (eds.) Language Sound Structure: Studies in Phonology. Cambridge: MIT Press. 107-136.* As for Jesperson, it would be [Chapter 13, Silbe, pp. 185-203](https://archive.org/details/lehrbuchderphone00jespuoft/page/184)

File: nltk/test/unit/test_tokenize.py
Comment (alvations) [inline nltk/test/unit/test_tokenize.py]: Lets call it `SyllableTokenizer` instead that'll be clearer.

File: nltk/tokenize/__init__.py
Comment (alvations) [inline nltk/tokenize/__init__.py]: How about `from nltk.tokenize.sonority_sequencing import SyllableTokenizer` ?

File: nltk/tokenize/ssp.py
Comment (alvations) [inline nltk/tokenize/ssp.py]: Rename this to `SyllableTokenizer`.

File: nltk/tokenize/ssp.py
Comment (alvations) [inline nltk/tokenize/ssp.py]: Lets specify the language here `def __init__(self, lang='en', snority_hierarchy=False):`

File: nltk/tokenize/ssp.py
Comment (alvations) [inline nltk/tokenize/ssp.py]: `if not sonority_hierarchy and lang='en':`

File: nltk/tokenize/ssp.py
Comment (alvations) [inline nltk/tokenize/ssp.py]: This might be slow if the list of syllable is long. Maybe a regex here would be faster.

File: nltk/tokenize/ssp.py
Comment (alvations) [inline nltk/tokenize/ssp.py]: Don't break lines in this manner, esp. when it's indicating string slice indices.

File: nltk/tokenize/ssp.py
Comment (alvations) [inline nltk/tokenize/ssp.py]: This is a good indication that looping through a trigrams of `self.syllables_values` would be good here.

**Review by alvations — CHANGES_REQUESTED** [review]
Happy New Year! 

Sorry for the delayed review. The class needs some refactoring and more importantly we might need to do some profiling and check whether the loops and if-else can be replace with more terse regex or ngrams operations to speed things up.

**Comment by henchc** [comment]
@alvations  I think I've addressed your comments, also added some more details to the information in the header. I'm not sure if there's a particular n-gram function you wanted me to use, so let me know if that section should be changed. Thanks!

File: nltk/tokenize/sonority_sequencing.py
Comment (alvations) [inline nltk/tokenize/sonority_sequencing.py]: What is the expect behavior for this exception? 

Raise and break the function?  Or silently let it go pass and perhaps raise a warning? 

If raising a warning, should the syllable values for the character that's not in the hierarchy be still added? Or leave it and skip the character (i.e. current behavior in the code)?

File: nltk/tokenize/sonority_sequencing.py
Comment (alvations) [inline nltk/tokenize/sonority_sequencing.py]: Hmmm, this isn't a good use of `self`. Since for every token the `syllables_values` resets and the content changes, it doesn't belong to the `SyllableTokenizer` variable. Try returning `syllable_values` instead.

```python
def assign_values(self, token):
    syllable_values = []
    ...
    return syllable_values
```

It should just be `syllables_values` here without the self. And then in the `tokenize()` function, it should be:

```python
syllables_values = self.assign_values(token)
```

**Comment by alvations** [comment]
@henchc I hope you don't mind that I took the liberty to change the `trigrams` iterations. 

I think now we just need to clarify the `try-except` in `assign_values()` and we should be clear to merge =)

File: nltk/tokenize/sonority_sequencing.py
Comment (henchc) [inline nltk/tokenize/sonority_sequencing.py:83]: add
```suggestion
self.phoneme_map[c.upper()] = len(sonority_hierarchy) - i
```
(see thought in PR thread)

**Comment by henchc** [comment]
Thanks, @alvations ! And thanks for cleaning up the n-gram function, looks much better :) .

Your question about the `try:` `except:` is a good one. I'm trying to think from the perspective of the most common use-case. I would think this is a user who just used `nltk.word_tokenize` and is sending the words through the syllabifier. 

1. First note in this thought exercise is that we should add the `.upper()` character set to each string in the current hierarchy so that upper case characters won't raise an error/warning. I'm thinking this could be added at [line 68](https://github.com/nltk/nltk/pull/2205/files#r246221923) since capitalization should not affect the sonority. The one issue I see is that the German `'ß'.upper()` goes to `"SS"` for some reason, but the single `"s"` should be in the original anyway so that will still cover `"S"` and the extra key shouldn't cause a problem.

2. Second note is that for an English use case directly after tokenization I imagine the main undefined characters being punctuation or foreign words with accents. For the most accurate return, I believe we should stop the function if it encounters an unknown character that is not punctuation (e.g. "déjà-vu"). Otherwise, I believe punctuation is safe to skip.

Does that make sense? I'm happy to make the changes if you agree, but I think we could go either way here and I'd defer to your sense for the library's standard practice.

**Comment by alvations** [comment]
Hmmm, lets try something like this before the loop in the `assign_value()`:

```python
import unicodedata 

token = 'déjà-vu'
unicodedata.normalize('NFKD', token.upper().lower()).encode('ascii', 'ignore').decode()

token='Vosßtraße'
unicodedata.normalize('NFKD', token.upper().lower()).encode('ascii', 'ignore').decode()
```

I wonder whether `str.encode('ascii', 'replace')` would be a better option. I'm not sure how NLTK users would use this so I think we should choose `'replace'` or `'ignore'` and just see how users react to the feature =)

-----

Meanwhile, please add the param documentation in the `___init__()` docstring:

```python
def __init__(...):
        """
        :param lang: Language parameter, default is English, 'en'
        :type lang: str
        :param sonority_hierarchy: Sonority hierarchy according to the Sonority Sequencing Principle.
        :type sonority_hierarchy: list(str)
        """
```

File: nltk/tokenize/sonority_sequencing.py
Comment (alvations) [reply nltk/tokenize/sonority_sequencing.py:83]: 👍 this should be done to handle the `ß` issue. Would it be better to:

```python
self.phoneme_map[c.upper().lower()] = len(sonority_hierarchy) - i
```

Comment (alvations) [reply nltk/tokenize/sonority_sequencing.py:83]: Hmmm but actually that complicates the access, you can loop through character by character after `ß -> ss` =)

**Comment by henchc** [comment]
I'm not sure `normalize` or `lower` is what we want, as that would transform the input. Why not let the input remain in its case with the special characters? I think we can just extend the hierarchy to include the `upper` form of the hierarchy, and warn about missing characters except punctuation? It could assign a default value of vowel and warn, so the user could add it to the hierarchy if they wanted.

**Comment by henchc** [comment]
I think my latest commit would be my suggested functionality, let me know what you think!

**Comment by no-identd** [comment]
German native speaker question:
Why remap `ß` to `ss` instead of `sz`?

**Comment by henchc** [comment]
Well we're not proposing the mapping it's just what python does with `"ß".upper()` . I was just calling that out as an edge case. I'm not sure why they would change it at all, especially since it messes with the `len`, but the "ß" as far as I know is to follow long vowels while "ss" follow short, though I think "ss" has begun to dominate beyond this context. "ẞ" was also used beyond that context before ("ich muss" now instead of "ich muß" was common before the Rechtschreibreform 1996). 

As for "sz", I think that was just orthographical when the "ß" first surfaced and the "sz" and "ss" sounds were converging in the language.

**Comment by henchc** [comment]
@alvations  any updates here?

**Comment by stevenbird** [comment]
@alvations are there any outstanding changes you have requested?

**Review by alvations — APPROVED** [review]

**Comment by alvations** [comment]
@henchc sorry for the really long delay, was busy with some other stuff. 

I've made some changes w.r.t. handling punctuations because previously the implementation chomps up the punctuations within the token and when the punctuation is a token itself, in the example:

```python
from nltk import word_tokenize
SSP = SyllableTokenizer()
text = "This is a foobar-like sentence."
print([SSP.tokenize(token) for token in word_tokenize(text)])
```

@henchc maybe we can handle the edge cases after the merge and see what users want to use the syllable tokenizer on =)

@stevenbird Once the test passes, it should be good to merge =)

**Comment by henchc** [comment]
No worries, and thanks, @alvations ! That makes sense to me, and yes, I'm happy to monitor and address edge cases, :)

**Comment by stevenbird** [comment]
Thanks @henchc and @alvations !

**Pull request merged**
